cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -411,26 +415,231 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
+   * This is not possible if the key is not present in the output schema, in which case the user will need to
+   * add it to the output schema. However, this is an unlikely scenario as most joins will want to
+   * preserve the key.
+   */
+  @VisibleForTesting
+  static List<Schema> deriveKeySchema(String joinerStageName, Map<String, List<String>> keys,
+                                      JoinDefinition joinDefinition) {
+    int numKeyFields = keys.values().iterator().next().size();
+
+    // (stage, field) -> JoinField
+    Table<String, String, JoinField> fieldTable = HashBasedTable.create();
+    for (JoinField joinField : joinDefinition.getSelectedFields()) {
+      fieldTable.put(joinField.getStageName(), joinField.getFieldName(), joinField);
+    }
+
+    /*
+        Suppose the join keys are:
+
+        A, (x, y)
+        B, (x, z))
+
+        and selected fields are:
+
+        A.x as id, A.y as name, B.z as item
+     */
+    Schema outputSchema = joinDefinition.getOutputSchema();
+    List<Schema> keySchema = new ArrayList<>(numKeyFields);
+    for (int i = 0; i < numKeyFields; i++) {
+      keySchema.add(null);
+    }
+    for (Map.Entry<String, List<String>> keyEntry : keys.entrySet()) {
+      String keyStage = keyEntry.getKey();
+      int keyFieldNum = 0;
+      for (String keyField : keyEntry.getValue()) {
+        // If key field is A.x, this will fetch (A.x as id)
+        // the JoinField might not exist. For example, B.x will not find anything in the output
+        JoinField selectedKeyField = fieldTable.get(keyStage, keyField);
+        if (selectedKeyField == null) {
+          continue;
+        }
+
+        // if the key field is A.x, JoinField is (A.x as id), and outputField is the 'id' field in the output schema
+        String outputFieldName = selectedKeyField.getAlias() == null ?
+          selectedKeyField.getFieldName() : selectedKeyField.getAlias();
+        Schema.Field outputField = outputSchema.getField(outputFieldName);
+
+        if (outputField == null) {
+          // this is an invalid join definition
+          throw new IllegalArgumentException(
+            String.format("Joiner stage '%s' provided an invalid definition. " +
+                            "The output schema does not contain a field for selected field '%s'.'%s'%s",
+                          joinerStageName, keyStage, selectedKeyField.getFieldName(),
+                          selectedKeyField.getAlias() == null ? "" : "as " + selectedKeyField.getAlias()));
+        }
+
+        // make the schema nullable because one stage might have it as non-nullable
+        // while another stage has it as nullable.
+        // for example, when joining on A.id = B.id,
+        // A.id might not be nullable, but B.id could be.
+        Schema keyFieldSchema = outputField.getSchema();
+        if (!keyFieldSchema.isNullable()) {
+          keyFieldSchema = Schema.nullableOf(keyFieldSchema);
```

Comment:
should we log something here as a TRACE or DEBUG message to let the user know we mark it as nullable, since there is possibility that it is not nullable

Reply:
this won't matter in the end because this is just intermediate data, the output will still use whatever the user provided. I can add a comment about that.